File: examples/sequential-cpp/example7.cpp
/* This code will generate a fractal image. Uses OpenCV, to compile:
   g++ example7.cpp `pkg-config --cflags --libs opencv`  */
   
#include <iostream>
#include <opencv/highgui.h>
#include "utils/cppheader.h"

using namespace std;

const int WIDTH = 1024;
const int HEIGHT = 768;
const int SCALEX = 1.500;
const int SCALEY = 1.500;

enum color {BLUE, GREEN, RED};

class Complex {
private:
	float real, img;
	
public:
	Complex(float r, float i) : real(r), img(i) {}
	
	float magnitude2() const {
		return (real * real) + (img * img);
	}
	
	Complex operator*(const Complex &a) {
		return Complex( ((real * a.real) - (img * a.img)),
		 				((img * a.real) + (real * a.img)) );
	}
	
	Complex operator+(const Complex &a) {
		return Complex( (real + a.real),
		 				(img + a.img) );
	}
};

class JuliaSet {
private:
	IplImage *img;
	
	int juliaValue(int x, int y, int width, int height) {
		int k;
		float jx = SCALEX * (float) (width / 2 - x) / (width / 2);
		float jy = SCALEY * (float) (height / 2 - y) / (height / 2);
		Complex c(-0.8, 0.156);
		Complex a(jx, jy);
	 
		for (k = 0; k < 200; k++) {
		    a = a*a + c;
		    if (a.magnitude2() > 1000) {
		        return 0;
		    }
		}
		return 1;
	}
	
public:
	JuliaSet(IplImage *image) : img(image) {}
	
	void doMagic() {
		int index, size, step;
		int ren, col, value;
		
		size = img->width * img->height;
		step = img->widthStep / sizeof(uchar);
		for (index = 0; index < size; index++) {
			ren = index / img->width;
			col = index % img->width;
			
			value = juliaValue(col, ren, img->width, img->height);
			
			img->imageData[(ren * step) + (col * img->nChannels) + RED] = (unsigned char) (255 * (0.4 * value));
			img->imageData[(ren * step) + (col * img->nChannels) + GREEN] = (unsigned char) (255 * (0.5 * value));
			img->imageData[(ren * step) + (col * img->nChannels) + BLUE] = (unsigned char) (255 * (0.7 * value));
		}
	}
};

int main(int argc, char* argv[]) {
	Timer t;
	double ms; 	
	IplImage* img=cvCreateImage(cvSize(WIDTH, HEIGHT), IPL_DEPTH_8U, 3);
	
	JuliaSet js(img);
	ms = 0;
	for (int i = 0; i < N; i++) {
		t.start();
		js.doMagic();
		ms += t.stop();
	}
	cout << "avg time = " << (ms /N) << endl;
	
	cvShowImage("CPU Julia | c(-0.8, 0.156)", img);
    cvWaitKey(0);
    cvDestroyWindow("CPU Julia | c(-0.8, 0.156)");
	
	return 0;
}

File: examples/sequential-java/Example7.java
/* This code will generate a fractal image. */
import java.awt.image.BufferedImage;

public class Example7 {
	private static final int WIDTH = 1024;
	private static final int HEIGHT = 768;
	private static final float SCALEX = 1.5f;
	private static final float SCALEY = 1.5f;
	private int array[], width, height;
	
	public Example7(int array[], int width, int height) {
		this.array = array;
		this.width = width;
		this.height = height;
	}
	
	private int juliaValue(int x, int y) {
		int k;
		float jx = SCALEX * (float) (width / 2 - x) / (width / 2);
		float jy = SCALEY * (float) (height / 2 - y) / (height / 2);
		Complex c = new Complex(-0.8f, 0.156f);
		Complex a = new Complex(jx, jy);
	 
		for (k = 0; k < 200; k++) {
		    a = (a.mult(a)).add(c);
		    if (a.magnitude2() > 1000) {
		        return 0;
		    }
		}
		return 1;
	}
	
	void doMagic() {
		int index, ren, col, value, pixel, r, g, b;
		
		for (index = 0; index < array.length; index++) {
			ren = index / width;
			col = index % width;
			pixel = array[index];
			
			value = juliaValue(col, ren);
			
			r = (int) (255 * (0.4 * value));
			g = (int) (255 * (0.5 * value));
			b = (int) (255 * (0.7 * value));
			
			array[index] =  (0xff000000)
							| (((int) r) << 16)
							| (((int) g) << 8)
							| (((int) b) << 0);
		}
	}
	
	public static void main(String args[]) {
		long startTime, stopTime;
		double acum = 0;
		
		int array[] = new int[WIDTH * HEIGHT];
		Example7 e = new Example7(array, WIDTH, HEIGHT);
		acum = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();
			e.doMagic();
			stopTime = System.currentTimeMillis();
			acum += (stopTime - startTime);
		}
		System.out.printf("avg time = %.5f\n", (acum / Utils.N));
		
		final BufferedImage bi = new BufferedImage(WIDTH, HEIGHT, BufferedImage.TYPE_INT_ARGB);
		bi.setRGB(0, 0, WIDTH, HEIGHT, array, 0, WIDTH);
		javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
               ImageFrame.showImage("CPU Julia | c(-0.8, 0.156)", bi);
            }
        });
	}
}
